contentgrid-spring-data-counting/src/main/java/com/contentgrid/spring/data/counting/ApproximatelyCountedPage.java

```diff
@@ -0,0 +1,107 @@
+package com.contentgrid.spring.data.counting;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import lombok.Getter;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.SliceImpl;
+
+public class ApproximatelyCountedPage<T> extends SliceImpl<T> implements Page<T> {
+
+    @Getter
+    private final CountResult countResult;
+
+    public ApproximatelyCountedPage(
+            List<T> content,
+            Pageable pageable,
+            boolean hasNext,
+            Supplier<Optional<CountResult>> countResult
+    ) {
+        super(content, pageable, hasNext);
+        this.countResult = adjustCountResult(calculateCountResult(countResult));
+    }
+
+    private ApproximatelyCountedPage(
+            List<T> content,
+            Pageable pageable,
+            boolean hasNext,
+            CountResult countResult
+    ) {
+        super(content, pageable, hasNext);
+        this.countResult = countResult;
+    }
+
+    @Override
+    public <U> ApproximatelyCountedPage<U> map(Function<? super T, ? extends U> converter) {
+        return new ApproximatelyCountedPage<>(
+                getConvertedContent(converter),
+                getPageable(),
+                hasNext(),
+                countResult
+        );
+    }
+
+    @Override
+    public int getTotalPages() {
+        if (getSize() == 0) {
+            return 1;
+        }
+        return (int) Math.ceil((double) getTotalElements() / getSize());
+    }
+
+    @Override
+    public long getTotalElements() {
+        return getCountResult().count();
+    }
+
+    @Override
+    public Pageable nextOrLastPageable() {
+        return super.nextOrLastPageable();
+    }
+
+    @Override
+    public Pageable previousOrFirstPageable() {
+        return super.previousOrFirstPageable();
+    }
```

Comment:
<s>
I was convinced that they are necessary because of the multiple inheritance diamond of death.

`ApproximatelyCountedPage` implements the `Slice` interface via 2 paths, and both paths have an implementation of `nextOrLastPageable()` (the default implementation in `Slice`).

```mermaid
classDiagram
    Slice <|-- SliceImpl
    Slice <|-- Page
    Page <|-- ApproximatelyCountedPage
    SliceImpl <|-- ApproximatelyCountedPage

    class Slice {
        +nextOrLastPageable()
    }
```

The actual overrides may change: `Page` or `SliceImpl` (or both) may provide their own default implementation, and as a subclass, we need to pick which one we want to invoke.
</s>

Looks like I can remove the methods, and it still compiles. I am quite surprised by that, because I was certain I added the overrides only because it wouldn't compile otherwise.